sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/FilterEstimation.scala

```diff
@@ -332,8 +332,41 @@ case class FilterEstimation(plan: Filter) extends Logging {
         colStatsMap.update(attr, newStats)
       }
 
-      Some(1.0 / BigDecimal(ndv))
-    } else {
+      if (colStat.histogram.isEmpty) {
+        // returns 1/ndv if there is no histogram
+        Some(1.0 / BigDecimal(ndv))
+      } else {
+        // We compute filter selectivity using Histogram information.
+        // Here we traverse histogram bins to locate the range of bins the literal values falls
+        // into.  For skewed distribution, a literal value can occupy multiple bins.
+        val hgmBins = colStat.histogram.get.bins
+        val datum = EstimationUtils.toDecimal(literal.value, literal.dataType).toDouble
```

Comment:
cc @wzhfy , you would refactor this part to always use Double for CBO computing?

Reply:
yes, I'll refactor this part.